### PR TITLE
8322692: ZGC: avoid over-unrolling due to hidden barrier size

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset refines the C2 loop unrolling heuristic by including an estimation of the final size of (Generational) ZGC barriers in the loop size computation. These are not exposed in C2's intermediate representation and thus currently ignored by the heuristic, which can lead to over-unrolling.

#### Testing

- tier1-5, stress test, fuzzing (windows-x64, linux-x64, linux-aarch64, macosx-x64, macosx-aarch64).
- tier6-9 (windows-x64, linux-x64, linux-aarch64, macosx-x64, macosx-aarch64, ZGC-specific tests only).

#### Performance and code size evaluation

- DaCapo, SPECjvm2008, SPECjbb2015 (linux-x64 with `-XX:+UseZGC -XX:+ZGenerational`). The changeset reduces slightly the size of the C2-generated code (around 0.3% fewer bytes per compiled bytecode for the DaCapo `fop` benchmark) and speeds up SPECjvm2008's `Serial` by around 4%.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322692](https://bugs.openjdk.org/browse/JDK-8322692): ZGC: avoid over-unrolling due to hidden barrier size (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17367/head:pull/17367` \
`$ git checkout pull/17367`

Update a local copy of the PR: \
`$ git checkout pull/17367` \
`$ git pull https://git.openjdk.org/jdk.git pull/17367/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17367`

View PR using the GUI difftool: \
`$ git pr show -t 17367`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17367.diff">https://git.openjdk.org/jdk/pull/17367.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17367#issuecomment-1888878964)

### CODE REVIEW

File: src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp [hunk 1]
```diff
@@ -320,6 +320,20 @@ void ZStoreBarrierStubC2::emit_code(MacroAssembler& masm) {
...
+  assert(barrier_data != 0, "should be a barrier node");
+  uint uncolor_or_color_size = node->is_Load() ? 1 : 2;
+  if ((barrier_data & ZBarrierElided) != 0) {
+    return uncolor_or_color_size;
+  }
+  // A compare and branch corresponds to approximately five Ideal nodes (Cmp,
+  // Bool, If, IfTrue, IfFalse). The runtime call corresponds to approximately
+  // seven more nodes (CallLeaf, control Proj, memory Proj, data Proj, Region,
+  // memory Phi, data Phi).
+  return uncolor_or_color_size + 12;
```
Comment (merykitty) [inline src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp:334]: I thought the runtime call does not lie inside the loop. Is it necessary to take them into account, too?

Comment (robcasloz) [reply src/hotspot/share/gc/z/c2/zBarrierSetC2.cpp:334]: Conceptually, the runtime call belongs to the loop, even if it is laid out in the cold section of the method. The current unrolling heuristic counts all basic blocks in the loop, regardless of whether they are hot or cold and how they are arranged in the final code. This changeset does the same for consistency.

**Comment by robcasloz** [comment]
The latest changes exclude the barrier slow path from the loop size estimation, as suggested by @fisk (offline) and @merykitty. Compared to the original changeset, this makes loop unrolling for ZGC more aggressive at the expense of code size, which is deemed acceptable in the typical scenarios in which ZGC is used. Compared to mainline, the code size improvement is now reduced to a mere 0.3% for DaCapo `fop` only, but in return SPECjvm2008 `Serial` is sped up by 4%. Please review.

**Review by fisk — APPROVED** [review]
Looks good. Thanks for fixing!

**Comment by robcasloz** [comment]
> Looks good. Thanks for fixing!

Thanks for reviewing, Erik!

File: src/hotspot/share/opto/loopTransform.cpp [hunk 2]
```diff
@@ -996,9 +998,12 @@ bool IdealLoopTree::policy_unroll(PhaseIdealLoop *phase) {
   uint body_size = _body.size();
   // Key test to unroll loop in CRC32 java code
   int xors_in_loop = 0;
-  // Also count ModL, DivL and MulL which expand mightly
+  // Also count ModL, DivL, MulL, and other nodes that expand mightly
   for (uint k = 0; k < _body.size(); k++) {
     Node* n = _body.at(k);
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/loopTransform.cpp:1003]: Yes, that sounds like a good idea for a later enhancement, please feel free to file a RFE. Note that the body size unit currently used by the heuristic is Ideal nodes (as opposed to machine instructions).

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by robcasloz** [comment]
> Looks good.

Thanks, Vladimir!

**Pull request closed without merging**